### PR TITLE
Query API and basic support on Metal

### LINKED ISSUE #1638: [RFC] Query resolve
## Overview

We have 3 types of queries in Vulkan: Occlusion, PipelineStatistics and Timestamp.
The first two are handled via `vkCmdBeginQuery`/`vkCmdEndQuery` and timestamps via `vkCmdWriteTimestamp`. Queries are allocated in heaps and marked as 'unavailable' at the beginning.
`vkCmdBeginQuery` calls activate queries and once the results can be obtained (after `vkCmdEndQuery`) the state will switch to `available`. `vkResetQueryPool` will reset them to `unavailable`.

## Mapping to Metal
Metal doesn't expose any significant kind of timestamp or pipeline statistics support (afaik) and only supports `Counting occlusion query`.
Therefore, `pipelineStatisticsQuery` feature should be exposed as `false`, `timestampComputeAndGraphics` as false and `timestampValidBits` as 0 for all queues.
Occlusion queries are configured at renderpass boundary. Binary occlusion support is required and counting optional. Results are directly written to a buffer.

## Mapping to D3D12

D3D12 uses query heaps, similar to Vulkan's query pools. The begin/end/write commands can be mapped without significant overhead to D3D12. Reset command implementation depends on the further Vulkan Portability API.
Resolving query data in D3D12 is implemented as a command list function `ResolveQueryData`(https://msdn.microsoft.com/de-de/library/windows/desktop/dn903896(v=vs.85).aspx). It write the current query values contiguously to a buffer (upload/download/local) as 64bit integers.
Vulkan exposes two functions to resolve queries: `vkGetQueryPoolResults` (device-side) and  `vkCmdCopyQueryPoolResults` (command buffer-side). Beside additional effort to submit a command buffer with a query resolve to emulate the device function, the `VkQueryResultFlags` are difficult to handle:

- `VK_QUERY_RESULT_64_BIT` denotes that values are written as u64 values (default: u32)
- `VK_QUERY_RESULT_WITH_AVAILABILITY_BIT` adds an additional integer value indicating if the specific query is available to the buffer.
- `VK_QUERY_RESULT_WAIT_BIT` waits on all queries to be available, else resulting in a possible device loss.

Additionally, users can define which pipeline statistics can should be recorded and only marked statistics get written to the buffer. D3D12 will emit all statistics.

#### Issues
* Track the status (`available/unavailable`) of each query in the query heap to correctly emit the availability flags if requested. As these happen async it needs to be done in the command buffer. For example, compute shader to write availability values into an hidden buffer on command buffer end if we have Begin/EndQuery commands.
* Compute shader to handle the case where `VK_QUERY_RESULT_64_BIT` is not set.
* vkGetQueryPoolResults requires additional synchronization to ensure that queries have been processed by the command queues when `VK_QUERY_RESULT_WAIT_BIT` is set.

I think these are the main issues so far. Metal requires a bit further analysis but shouldn't cause major complications.

### PR DESCRIPTION
**Pull request by kvark**
Fixes #1638 (for the most part)
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [ ] tested examples with the following backends:
- [ ] `rustfmt` run on changed code

### CODE REVIEW

**Comment by kvark** [comment]
Looks like we are blocked on https://github.com/MaikKlein/ash/issues/100 for Vulkan :(
cc @MaikKlein

**Comment by kvark** [comment]
Current status: 136 / 150 occlusion tests passing
Known 2 issues:
  1. sometimes getting the query results synchronously gets stuck
  2. in some cases copying results with availability doesn't produce correct results

**Comment by kvark** [comment]
Vulkan is blocked on https://github.com/MaikKlein/ash/issues/105 now

**Comment by kvark** [comment]
Update on Metal: all the known issues are resolved. We pass 150/150 tests without failures or hangs :tada: 
The PR is blocked on reviews and getting Vulkan backend to compile. In the worst case we'll comment out the Vulkan parts that are blocking.

File: src/backend/vulkan/src/command.rs [hunk 1]
```diff
@@ -849,10 +849,10 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
     fn begin_query(
         &mut self,
         query: query::Query<Backend>,
-        control: query::QueryControl,
+        control: query::ControlFlags,
     ) {
         let mut flags = vk::QueryControlFlags::empty();
-        if control.contains(query::QueryControl::PRECISE) {
+        if control.contains(query::ControlFlags::PRECISE) {
```
Comment (grovesNL) [inline src/backend/vulkan/src/command.rs]: `map_query_control_flags` instead?

File: src/backend/metal/src/device.rs [hunk 2]
```diff
@@ -2291,12 +2304,102 @@ impl hal::Device<Backend> for Device {
...
+        data: &mut [u8], stride: buffer::Offset,
+        flags: query::ResultFlags,
+    ) -> Result<bool, query::Error> {
+        let is_ready = match *pool {
+            native::QueryPool::Occlusion(ref pool_range) => {
+                let visibility = &self.shared.visibility;
+                let is_ready = if flags.contains(query::ResultFlags::WAIT) {
+                    let mut guard = visibility.allocator.lock();
+                    while !visibility.are_available(pool_range.start, &queries) {
+                        println!("query: not available"); //TEMP
```
Comment (grovesNL) [inline src/backend/metal/src/device.rs]: Is this left here intentionally?

File: src/backend/metal/src/command.rs [hunk 3]
```diff
@@ -1656,13 +1682,21 @@ impl RawCommandQueue<Backend> for CommandQueue {
...
                 let free_textures = self.retained_textures
                     .drain(..)
                     .collect::<SmallVec<[_; BLOCK_BUCKET]>>();
+                let visibility = if self.active_visibility_queries.is_empty() {
+                    None
+                } else {
+                    let queries = self.active_visibility_queries
+                        .drain(..)
+                        .collect::<SmallVec<[_; BLOCK_BUCKET]>>();
+                    Some((Arc::clone(&self.shared), queries))
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: Why do we need to clone`self.shared`?

File: src/backend/metal/src/command.rs [hunk 4]
```diff
@@ -3454,31 +3501,192 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
+                            let com_avail = soft::BlitCommand::CopyBuffer {
+                                src: AsNative::from(visibility.buffer.as_ref()),
+                                dst: AsNative::from(buffer.raw.as_ref()),
+                                region: com::BufferCopy {
+                                    src: visibility.availability_offset + absolute_index * size_meta,
+                                    dst: dst_offset + size_payload,
+                                    size: size_meta,
+                                },
+                            };
+                            // An extra paddig is requred if the client expects 64bits availability without a wait
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: `paddig` -> `padding`

File: src/backend/metal/src/command.rs (hunk 3 above)
Comment (grovesNL) [reply src/backend/metal/src/command.rs]: Oh, I didn't notice the callback block. Sounds good

File: src/backend/metal/src/command.rs [hunk 5]
```diff
@@ -3454,31 +3501,192 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
+                            region: com::BufferCopy {
+                                src: absolute_index * size_data,
+                                dst: dst_offset,
+                                size: size_payload,
+                            },
+                        };
+
+                        let (com_avail, com_pad) = if flags.contains(query::ResultFlags::WITH_AVAILABILITY | query::ResultFlags::WAIT) {
+                            // Technically waiting is a no-op on a single queue. However,
+                            // the client expects the avaiability to be set regardless.
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: `avaiability` -> `availability`

**Comment by kvark** [comment]
Landing, as per approval on gitter.
bors r=grovesNL

**Comment by kvark** [comment]
Bors retry

**Comment by kvark** [comment]
Bors r=grovesNL

**Comment by kvark** [comment]
Oh yes you are, silly bors!

**Pull request merged**